zebra-state/src/service/non_finalized_state/tests/prop.rs

```diff
@@ -59,6 +74,160 @@ fn finalized_equals_pushed() -> Result<()> {
         }
 
         prop_assert_eq!(full_chain.blocks.len(), partial_chain.blocks.len());
+        prop_assert!(full_chain.eq_internal_state(&partial_chain));
+    });
+
+    Ok(())
+}
+
+/// Check that rejected blocks do not change the internal state of a chain
+/// in a non-finalized state.
+#[test]
+fn rejection_restores_internal_state() -> Result<()> {
+    zebra_test::init();
+
+    proptest!(ProptestConfig::with_cases(env::var("PROPTEST_CASES")
+                                         .ok()
+                                         .and_then(|v| v.parse().ok())
+                                         .unwrap_or(DEFAULT_PARTIAL_CHAIN_PROPTEST_CASES)),
+              |((chain, valid_count, network, mut bad_block) in (PreparedChain::default(), any::<bool>(), any::<bool>())
+                .prop_flat_map(|((chain, valid_count, network), is_nu5, is_v5)| {
+                    let next_height = chain[valid_count - 1].height;
+                    (
+                        Just(chain),
+                        Just(valid_count),
+                        Just(network),
+                        // generate a Canopy or NU5 block with v4 or v5 transactions
+                        LedgerState::height_strategy(
+                            next_height,
+                            if is_nu5 { Nu5 } else { Canopy },
+                            if is_nu5 && is_v5 { 5 } else { 4 },
+                            true,
+                        )
+                            .prop_flat_map(Block::arbitrary_with)
+                            .prop_map(DisplayToDebug)
+                    )
+                }
+                ))| {
+                  let mut state = NonFinalizedState::new(network);
+
+                  // use `valid_count` as the number of valid blocks before an invalid block
+                  let valid_tip_height = chain[valid_count - 1].height;
+                  let valid_tip_hash = chain[valid_count - 1].hash;
+                  let chain = chain.iter().take(valid_count);
+
+                  prop_assert!(state.eq_internal_state(&state));
+
+                  for (index, block) in chain.enumerate() {
+                      if index == 0 {
+                          state.commit_new_chain(block.clone())?;
+                      } else {
+                          state.commit_block(block.clone())?;
+                      }
+                      prop_assert!(state.eq_internal_state(&state));
+                  }
```

Comment:
Idea: just a refactor idea, but I'm not sure if it actually improves the readability.

```suggestion
                  let mut chain = chain.iter().take(valid_count).cloned();

                  prop_assert!(state.eq_internal_state(&state));

                  if let Some(first_block) = chain.next() {
                      state.commit_new_chain(block)?;
                      prop_assert!(state.eq_internal_state(&state));
                  }

                  for block in chain {
                      state.commit_block(block)?;
                      prop_assert!(state.eq_internal_state(&state));
                  }
```